### PR TITLE
rustc: Add a `#[wasm_custom_section]` attribute

### PR DESCRIPTION
**Pull request by alexcrichton**
This commit is an implementation of adding custom sections to wasm artifacts in
rustc. The intention here is to expose the ability of the wasm binary format to
contain custom sections with arbitrary user-defined data. Currently neither our
version of LLVM nor LLD supports this so the implementation is currently custom
to rustc itself.

The implementation here is to attach a `#[wasm_custom_section = "foo"]`
attribute to any `const` which has a type like `[u8; N]`. Other types of
constants aren't supported yet but may be added one day! This should hopefully
be enough to get off the ground with *some* custom section support.

The current semantics are that any constant tagged with `#[wasm_custom_section]`
section will be *appended* to the corresponding section in the final output wasm
artifact (and this affects dependencies linked in as well, not just the final
crate). This means that whatever is interpreting the contents must be able to
interpret binary-concatenated sections (or each constant needs to be in its own
custom section).

To test this change the existing `run-make` test suite was moved to a
`run-make-fulldeps` folder and a new `run-make` test suite was added which
applies to all targets by default. This test suite currently only has one test
which only runs for the wasm target (using a node.js script to use `WebAssembly`
in JS to parse the wasm output).

### CODE REVIEW

File: src/tools/compiletest/src/runtest.rs [hunk 1]
```diff
@@ -2359,11 +2359,6 @@ impl<'test> TestCx<'test> {
     }
 
     fn run_rmake_test(&self) {
-        // FIXME(#11094): we should fix these tests
-        if self.config.host != self.config.target {
```
Comment (Mark-Simulacrum) [inline src/tools/compiletest/src/runtest.rs]: hm, won't this make us run more tests now? Or do we handle this elsewhere for fulldeps suites somehow?

Comment (alexcrichton) [reply src/tools/compiletest/src/runtest.rs]: I think this was added long long ago when we still ran more tests in more situations (like tests on dist builders), but nowadays I don't think this is ever triggered. If I'm wrong though this'll just fail in CI

File: src/librustc_trans/back/wasm.rs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+use std::fs;
+use std::path::Path;
+use std::collections::BTreeMap;
+
+use serialize::leb128;
+
+pub fn add_custom_sections(path: &Path, sections: &BTreeMap<String, Vec<u8>>) {
+    let mut wasm = fs::read(path).expect("failed to read wasm output");
+
+    // see https://github.com/WebAssembly/design/blob/master/BinaryEncoding.md#module-structure
```
Comment (pepyakin) [inline src/librustc_trans/back/wasm.rs]: Can we really use https://webassembly.github.io/spec/core/binary/modules.html#custom-section for the reference?

**Comment by nikomatsakis** [comment]
[Travis is sad:](https://travis-ci.org/rust-lang/rust/builds/353397570#L1349)

you need to replace `ty::ParamEnv::empty(traits::Reveal::All)` with `ty::ParamEnv::reveal_all()`

File: src/libsyntax/feature_gate.rs [hunk 3]
```diff
@@ -455,6 +455,9 @@ declare_features! (
 
     // Parentheses in patterns
     (active, pattern_parentheses, "1.26.0", None, None),
+
+    // The #[wasm_custom_section] attribute
+    (active, wasm_custom_section, "1.26.0", None, None),
```
Comment (nikomatsakis) [inline src/libsyntax/feature_gate.rs]: (don't we need a feature-gate test for this ?)

**Review by nikomatsakis — APPROVED** [review]
r=me once travis is happy

File: src/libsyntax/feature_gate.rs (hunk 3 above)
Comment (alexcrichton) [reply src/libsyntax/feature_gate.rs]: Heh probably!

**Comment by kennytm** [comment]
cargotest ICE'd

<details>

```
   Compiling selectors v0.19.0 (file:///C:/projects/rust/build/ct/servo/components/selectors)
   Compiling euclid v0.15.5
thread 'rustc' panicked at 'index out of bounds: the len is 9 but the index is 136', C:\projects\rust\src\libcore\slice\mod.rs:865:10
note: Run with `RUST_BACKTRACE=1` for a backtrace.
error: internal compiler error: unexpected panic
note: the compiler unexpectedly panicked. this is a bug.
note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports
note: rustc 1.26.0-dev running on x86_64-pc-windows-msvc
note: compiler flags: -C codegen-units=4 -C debuginfo=2 -C incremental --crate-type bin
note: some of the compiler flags provided by cargo are hidden
error: Could not compile `selectors`.
warning: build failed, waiting for other jobs to finish...
error: build failed
thread 'main' panicked at 'tests failed for https://github.com/servo/servo', tools\cargotest\main.rs:100:9
note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

</details>

**Comment by alexcrichton** [comment]
@bors: r-

nope bors, this should not be approved

**Comment by eddyb** [comment]
Do we want to use `#[section_name]` for this long-term by any chance?

**Comment by alexcrichton** [comment]
@eddyb perhaps, although it has a pretty radically different meaning on wasm so far which doens't map well (aka we bypass llvm completely)

**Pull request merged**